bindings/python/pymongoarrow/lib.pyx

```diff
@@ -257,20 +270,25 @@ def process_bson_stream(bson_stream, context, arr_value_builder=None):
                         builder.append_null()
                 elif ftype == t_document:
                     if value_t == BSON_TYPE_DOCUMENT:
-                        bson_iter_document(&doc_iter, &doc_buf_len, &doc_buf)
-                        if doc_buf_len <= 0:
+                        bson_iter_document(&doc_iter, &val_buf_len, &val_buf)
+                        if val_buf_len <= 0:
                             raise ValueError("Subdocument is invalid")
-                        builder.append(<bytes>doc_buf[:doc_buf_len])
+                        builder.append(<bytes>val_buf[:val_buf_len])
                     else:
                         builder.append_null()
                 elif ftype == t_array:
                     if value_t == BSON_TYPE_ARRAY:
-                        bson_iter_array(&doc_iter, &doc_buf_len, &doc_buf)
-                        if doc_buf_len <= 0:
+                        bson_iter_array(&doc_iter, &val_buf_len, &val_buf)
+                        if val_buf_len <= 0:
                             raise ValueError("Subarray is invalid")
-                        builder.append(<bytes>doc_buf[:doc_buf_len])
+                        builder.append(<bytes>val_buf[:val_buf_len])
                     else:
                         builder.append_null()
+                elif ftype == t_binary:
+                    if value_t == BSON_TYPE_BINARY:
+                        bson_iter_binary (&doc_iter, &subtype,
+                            &val_buf_len, &val_buf)
+                        builder.append(<bytes>val_buf[:val_buf_len])
```

Comment:
Shouldn't we validate that the subtype is the expected one? Could you add a test for reading data with mixed subtypes?

Reply:
Done and done